s3transfer/tasks.py

```diff
@@ -279,6 +284,20 @@ def _submit(self, transfer_future, **kwargs):
         """
         raise NotImplementedError('_submit() must be implemented')
 
+    def _conditionally_execute_main(self, kwargs):
+        # If the task is not done (really only if some other related
+        # task to the TransferFuture had failed) then execute the task's
+        # main() method.
+        if not self._transfer_coordinator.done():
+            return self._execute_main(kwargs)
+        # If the transfer coordinator signals it is done prior to the
+        # submission task running _main, the submission task will not
+        # submit any further tasks and therefore must take responsibility
+        # to announce done because there will be no tasks submitted to
+        # announce done for the transfer.
+        else:
+            self._transfer_coordinator.announce_done()
```

Comment:
Eh I think ``done()`` is a poorly chosen name. Mainly because ``done()`` does not actually mean the transfer is done, as there is nothing left to do. ``done()`` just means that the status has been set to a terminal status, but there could be extra activities happening. The **only** thing that indicates there is nothing left to do in the transfer is the setting of that done event. The main problem is that status and this done event are set independently. In this case, I am using the fact that these are independently set to pretty much determine if it is in a "is shutting down" state.

I think given the time, I would do the following to make this logic more clear:

1) Introduce a "shutting-down" state that if an error or cancel occurred, it would be set and then shifted over to failed and cancelled once done was announced.
2) Move "success" state over from "running" when done is announced, instead of setting it when ``set_result()`` is called.
3) Remove this logic to execute main if it is not shutting down.
4) Abstract all this logic more because it is getting complicated.

Does that make sense? I could make this change but seems outside the scope of this PR, but I'd be happy to make that change. What are your thoughts/suggestions?